pageserver/src/repository.rs

```diff
@@ -116,15 +124,37 @@ pub trait Timeline {
                 == pg_constants::XLOG_SMGR_TRUNCATE
         {
             let truncate = XlSmgrTruncate::decode(&decoded);
+            let mut rel = RelTag {
+                spcnode: truncate.rnode.spcnode,
+                dbnode: truncate.rnode.dbnode,
+                relnode: truncate.rnode.relnode,
+                forknum: pg_constants::MAIN_FORKNUM,
+            };
             if (truncate.flags & pg_constants::SMGR_TRUNCATE_HEAP) != 0 {
-                let rel = RelTag {
-                    spcnode: truncate.rnode.spcnode,
-                    dbnode: truncate.rnode.dbnode,
-                    relnode: truncate.rnode.relnode,
-                    forknum: pg_constants::MAIN_FORKNUM,
-                };
                 self.put_truncation(rel, lsn, truncate.blkno)?;
             }
+            if (truncate.flags & pg_constants::SMGR_TRUNCATE_FSM) != 0 {
+                if truncate.blkno == 0 {
+                    rel.forknum = pg_constants::FSM_FORKNUM;
+                    self.put_truncation(rel, lsn, truncate.blkno)?;
+                } else {
+                    // TODO: handle partial truncation of FSM:
+                    // need to map heap block number to FSM block number
+                    // and clear bits in the tail block
+                    info!("Partial truncation of FSM is not supported");
+                }
+            }
+            if (truncate.flags & pg_constants::SMGR_TRUNCATE_VM) != 0 {
+                if truncate.blkno == 0 {
+                    rel.forknum = pg_constants::VISIBILITYMAP_FORKNUM;
```

Comment:
It would be nice if `XlSmgrTruncate.flags` had a native representation, instead of just doing raw bit-math everywhere.

Reply:
Sorry, do not understand you. "flags" field in xl_truncate structure (taken from Postgres code) is  really used as bitmask.

Reply:
I am suggesting two things:
1. The rust data structure should not allow invalid combinations of flags.
2. We should hide the bit-checking inside the rust data structure, allowing e.g. `truncate.flags.is_heap()`

This is not specific to this particular field, just how I would prefer to handle bitfields in general. There may be crates that will help: `bitflags`, for example.

Unless they turn out to be mutually exclusive, then I suggest an enum.

Reply:
Sorry, I still do not understand your arguments. There are three flags and any combination of them is valid. I do not see that encapsulating this field in separate class and providing methods like is_heap() is really good idea in this case. From my point of view it just adds extra code without increasing readability or safety. May be it is just my own aberration, caused by more than 30-year C experience. I do not see anything criminal or magic in bit manipulation operations. From my point of view adding methods (or macros in C) like is_heap() is useful when this check is performed many times. In this case construction with bit AND and really looks too bulky. But if it is used just in one place, then adding all wrappers around this bitflags is overkill from my point of view.

Reply:
There is no type safety in the expression `flags & SMGR_TRUNCATE_VM`.

If `flags` had a proper type, the compiler would be able to enforce that that bit is appropriate for this field, but at present there is no checking at all: The compiler would also allow `flags & XACT_XINFO_HAS_INVALS`, which is probably nonsense.

Maybe if it's only used once it's not worth bothering. But `pg_constants.rs` keeps growing, and there is no type safety anywhere. I don't see any comments that tell the reader where it's safe to use `SGMR_TRUNCATE_*` values.

Reply:
Perhaps [bitflags](https://docs.rs/bitflags/1.2.1/bitflags/) could give us type safety.